File: 0201-0300/216-combination-sum-iii/216-combination-sum-iii.js
// 216. Combination Sum III
// https://leetcode.com/problems/combination-sum-iii/description/

/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
var combinationSum3 = function(k, n) {
  function generateCombination(count, pos, c, n) {
    if (count === k) {
      if (n === 0) res.push([...c]);
    } else {
      for (let i = pos; i < 10; i++) {
        c.push(i);
        generateCombination(count+1, i+1, c, n - i);
        c.pop(i);
      }
    }
  }

  const res = [];
  generateCombination(0, 1, [], n);
  return res;
};

console.info(combinationSum3(3, 7));
console.info(combinationSum3(3, 9));

File: 0201-0300/216-combination-sum-iii/216-combination-sum-iii.cc
class Solution {
    vector<vector<int>> res;
    vector<int> nums;

    void backtracking(int start, int &sum, int k, int n) {
        if (nums.size() == k && sum == n) {
            res.push_back(nums);
            return;
        }

        for (int i = start; i <= n - sum && i <= 9; ++i) {
            nums.push_back(i);
            sum += i;
            backtracking(i+1, sum, k, n);
            sum -= i;
            nums.pop_back();
        }
    }

public:
    vector<vector<int>> combinationSum3(int k, int n) {
        int sum = 0;
        backtracking(1, sum, k, n);
        return res;
    }
};
